### PR TITLE
feat:  add groovy support for jbang

### LINKED ISSUE #1113: Groovy support?
Any plan to support Groovy alike Kotlin?    `jbang hello.groovy` 

```
//GROOVY 3.0.9

println("Hello Groovy")
```

I have some Ops scripts written by Groovy, and I would like to run these groovy scripts with jbang support.  Auto JDK, alias, jbang-catalog.json features are very useful to run Groovy scripts.

### PR DESCRIPTION
**Pull request by linux-china**
Fix #1113

### CODE REVIEW

**Comment by maxandersen** [comment]
awesome - didn't expect you to do it that fast :)

Could you add some groovy based tests in this PR?

File: src/main/java/dev/jbang/net/GroovyManager.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	public static Path getGroovyPath(String version) {
+		return getGroovycsPath().resolve(version);
+	}
+
+	private static Path getGroovycsPath() {
+		return Settings.getCacheDir(Cache.CacheClass.groovycs);
+	}
+
+	private static String getGroovyDownloadUrl(String version) {
+		if (version.startsWith("4.0")) {
```
Comment (aalmiray) [inline src/main/java/dev/jbang/net/GroovyManager.java:78]: this will break the moment Groovy releases a version greater than `4.0.x`. Perhaps using a regex such as "(\\d+)\\..+", then parse as int the first matching group and check if it's >= 4

File: src/main/java/dev/jbang/net/GroovyManager.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			if (!Files.isDirectory(groovyDir) && Files.isDirectory(groovyOldDir)) {
+				try {
+					Files.move(groovyOldDir, groovyDir);
+				} catch (IOException ex) {
+					// Ignore
+				}
+			}
+			Util.errorMsg("Required Groovy version not possible to download or install.");
+			throw new ExitException(EXIT_UNEXPECTED_STATE,
+					"Unable to download or install groovyc version " + version, e);
```
Comment (aalmiray) [inline src/main/java/dev/jbang/net/GroovyManager.java:65]: nitpick, change `groovyc` to just `groovy`.

File: src/main/java/dev/jbang/util/Util.java [hunk 3]
```diff
@@ -72,7 +72,7 @@
 	public static final Pattern mainClassMethod = Pattern.compile(
 			"(?<=\\n|\\A)(?:public\\s)\\s*(class)\\s*([^\\n\\s]*)");
 
-	private static final List<String> EXTENSIONS = asList(".java", ".jsh", ".kt");
+	private static final List<String> EXTENSIONS = asList(".java", ".jsh", ".kts", ".groovy");
```
Comment (aalmiray) [inline src/main/java/dev/jbang/util/Util.java:75]: This are 2 unrelated changes.

File: src/main/java/dev/jbang/source/GroovyScriptSource.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	public Predicate<ClassInfo> getMainFinder() {
+		return pubClass -> pubClass.method("main", BaseBuildCommand.STRINGARRAYTYPE) != null
+				|| pubClass.method("main") != null;
+	}
+
+	@Override
+	public List<String> getAllDependencies() {
+		final List<String> allDependencies = super.getAllDependencies();
+		final String groovyVersion = getGroovyVersion();
+		if (groovyVersion.startsWith("4.")) {
```
Comment (aalmiray) [inline src/main/java/dev/jbang/source/GroovyScriptSource.java:41]: To future proof this condition I'd recommend grabbing the first number from the version, parse it as an int an check if it's >= 4

File: src/main/java/dev/jbang/util/Util.java (hunk 3 above)
Comment (maxandersen) [reply src/main/java/dev/jbang/util/Util.java:75]: Kts!= kt. 

Kt is compiled first. Kts is not. 

Jbang does not support kts style scripts. At least not in its current shape.

Comment (linux-china) [reply src/main/java/dev/jbang/util/Util.java:75]: yes, it's my fault.  I'm working on kts support for jbang, and sorry for a mistake introduced.

Comment (aalmiray) [reply src/main/java/dev/jbang/util/Util.java:75]: It's OK. You just have to keep the changes separate from each other 😄

File: src/main/java/dev/jbang/source/GroovyScriptSource.java (hunk 4 above)
Comment (linux-china) [reply src/main/java/dev/jbang/source/GroovyScriptSource.java:41]: groovyMajorVersion added.

```java
                 int groovyMajorVersion = 3;
		Matcher matcher = Pattern.compile("(\\d)\\..*").matcher(version);
		if (matcher.find()) {
			groovyMajorVersion = Integer.parseInt(matcher.group(1));
		}
```

File: src/main/java/dev/jbang/net/GroovyManager.java (hunk 2 above)
Comment (linux-china) [reply src/main/java/dev/jbang/net/GroovyManager.java]: Fixed

**Comment by linux-china** [comment]
> Could you add some groovy based tests in this PR?

Add itests/helloworld.groovy with Java 11 API.  `jbang itests/helloworld.groovy`

**Comment by maxandersen** [comment]
is there some api like we have for jdk with foojay to get matching kotlin/groovy sdks without requiring users to expliclitly state exact versions ?

**Comment by aalmiray** [comment]
I don't think there is such API for Groovy. We can ask at the Groovy ML though.

**Comment by paulk-asert** [comment]
> is there some api like we have for jdk with foojay to get matching kotlin/groovy sdks without requiring users to expliclitly state exact versions ?

Can you point me to the foojay api?

**Comment by linux-china** [comment]
> > is there some api like we have for jdk with foojay to get matching kotlin/groovy sdks without requiring users to expliclitly state exact versions ?
> 
> Can you point me to the foojay api?

Now foojay does not support SDK download for kotlin, groovy, scala etc.  https://api.foojay.io/swagger-ui

I can parse maven-metadata.xml to get exact version for  major  + minor version,  and is it ok?

**Comment by aalmiray** [comment]
@paulk-asert you can find it at https://github.com/foojayio/discoapi
@hansolo takes care of it.

File: src/main/java/dev/jbang/net/GroovyManager.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	private static String getGroovyDownloadUrl(String version) {
+		int groovyMajorVersion = 3;
+		Matcher matcher = Pattern.compile("(\\d)\\..*").matcher(version);
+		if (matcher.find()) {
+			groovyMajorVersion = Integer.parseInt(matcher.group(1));
+		}
+		if (groovyMajorVersion >= 4) {
+			return String.format(
```
Comment (maxandersen) [inline src/main/java/dev/jbang/net/GroovyManager.java:86]: if these are artifacts shouldn't we just fetch using the maven resolver ?

File: src/main/java/dev/jbang/source/GroovyScriptSource.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+			allDependencies.add("org.apache.groovy:groovy:" + groovyVersion);
```
Comment (maxandersen) [inline src/main/java/dev/jbang/source/GroovyScriptSource.java:42]: why is this one added ? is it as a "should always be there" kind of dependency?

Comment (linux-china) [reply src/main/java/dev/jbang/source/GroovyScriptSource.java:42]: Yes, if you want to run Java class compiled by groovyc, and you should add groovy-version.jar to include Groovy APIs
and Groovy Development Kit enhancements. 

```
❯ javap Hello
Compiled from "Hello.groovy"
public class Hello extends groovy.lang.Script {
  public static transient boolean __$stMC;
  public Hello();
  public Hello(groovy.lang.Binding);
  public static void main(java.lang.String...);
  public java.lang.Object run();
  protected groovy.lang.MetaClass $getStaticMetaClass();
}
```

File: src/main/java/dev/jbang/net/GroovyManager.java (hunk 5 above)
Comment (linux-china) [reply src/main/java/dev/jbang/net/GroovyManager.java:86]: you mean org.codehaus.groovy:groovy-binary:3.0.9?

**Comment by linux-china** [comment]
Could we make it happened before Groovy 4.0 released?  :)

File: src/main/java/dev/jbang/net/GroovyManager.java (hunk 5 above)
Comment (maxandersen) [reply src/main/java/dev/jbang/net/GroovyManager.java:86]: Yes?

**Review by maxandersen — CHANGES_REQUESTED** [review]

**Comment by maxandersen** [comment]
fyi, This needs a rebase to merge cleanly.

**Comment by maxandersen** [comment]
tried `jbang  https://gitlab.com/-/snippets/2224724/raw/main/send.groovy` which fails downloading the groovy all...

```
[jbang] System Java version detected as 11
[jbang] System Java version matches requested version 11
[jbang] Resolving dependencies...
[jbang]     Resolving com.puravida-software.groogle:groogle-gmail:jar:3.0.0...[jbang] Deleting folder /Users/max/.jbang/cache/jars/send.groovy.java.ac022e7681e0deda972373b8805b17735daaf6cd7b704d8a83a4d31d774abcff.jar.tmp
[jbang] [ERROR] Could not resolve dependency
dev.jbang.cli.ExitException: Could not resolve dependency
	at dev.jbang.dependencies.DependencyUtil.resolveDependenciesViaAether(DependencyUtil.java:202)
	at dev.jbang.dependencies.DependencyUtil.resolveDependencies(DependencyUtil.java:113)
	at dev.jbang.dependencies.DependencyUtil.resolveDependencies(DependencyUtil.java:62)
	at dev.jbang.dependencies.DependencyResolver.resolve(DependencyResolver.java:67)
	at dev.jbang.source.RunContext.resolveClassPath(RunContext.java:305)
	at dev.jbang.cli.BaseBuildCommand.buildJar(BaseBuildCommand.java:152)
	at dev.jbang.cli.BaseBuildCommand.build(BaseBuildCommand.java:127)
	at dev.jbang.cli.BaseBuildCommand.buildIfNeeded(BaseBuildCommand.java:91)
	at dev.jbang.cli.Run.prepareArtifacts(Run.java:100)
	at dev.jbang.cli.Run.doCall(Run.java:76)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:85)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:12)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at dev.jbang.cli.JBang$3.handle(JBang.java:135)
	at dev.jbang.cli.JBang$3.handle(JBang.java:130)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at dev.jbang.Main.main(Main.java:14)
Caused by: org.jboss.shrinkwrap.resolver.api.NoResolvedResultException: Unable to collect/resolve dependency tree for a resolution due to: Could not find artifact org.codehaus.groovy:groovy-all:jar:3.0.9 in mavencentral (https://repo1.maven.org/maven2/), caused by: Could not find artifact org.codehaus.groovy:groovy-all:jar:3.0.9 in mavencentral (https://repo1.maven.org/maven2/)
	at org.jboss.shrinkwrap.resolver.impl.maven.MavenWorkingSessionImpl.wrapException(MavenWorkingSessionImpl.java:503)
	at org.jboss.shrinkwrap.resolver.impl.maven.MavenWorkingSessionImpl.resolveDependencies(MavenWorkingSessionImpl.java:242)
	at org.jboss.shrinkwrap.resolver.impl.maven.MavenStrategyStageBaseImpl.using(MavenStrategyStageBaseImpl.java:70)
	at org.jboss.shrinkwrap.resolver.impl.maven.MavenStrategyStageBaseImpl.withTransitivity(MavenStrategyStageBaseImpl.java:52)
	at org.jboss.shrinkwrap.resolver.impl.maven.MavenStrategyStageBaseImpl.withTransitivity(MavenStrategyStageBaseImpl.java:39)
	at dev.jbang.dependencies.DependencyUtil.resolveDependenciesViaAether(DependencyUtil.java:192)
	... 20 more
Caused by: org.eclipse.aether.resolution.DependencyResolutionException: Could not find artifact org.codehaus.groovy:groovy-all:jar:3.0.9 in mavencentral (https://repo1.maven.org/maven2/)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:357)
	at org.jboss.shrinkwrap.resolver.impl.maven.bootstrap.MavenRepositorySystem.resolveDependencies(MavenRepositorySystem.java:121)
	at org.jboss.shrinkwrap.resolver.impl.maven.MavenWorkingSessionImpl.resolveDependencies(MavenWorkingSessionImpl.java:239)
	... 24 more
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact org.codehaus.groovy:groovy-all:jar:3.0.9 in mavencentral (https://repo1.maven.org/maven2/)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:424)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:229)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:340)
	... 26 more
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.codehaus.groovy:groovy-all:jar:3.0.9 in mavencentral (https://repo1.maven.org/maven2/)
	at org.eclipse.aether.connector.basic.ArtifactTransportListener.transferFailed(ArtifactTransportListener.java:48)
	at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:369)
	at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:75)
	at org.eclipse.aether.connector.basic.BasicRepositoryConnector$DirectExecutor.execute(BasicRepositoryConnector.java:644)
	at org.eclipse.aether.connector.basic.BasicRepositoryConnector.get(BasicRepositoryConnector.java:262)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.performDownloads(DefaultArtifactResolver.java:499)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:401)
	... 28 more
[jbang] If you believe this a bug in jbang, open an issue at https://github.com/jbangdev/jbang/issues
```

shouldn't that not just work?

**Comment by maxandersen** [comment]
like wise this is failing too:

```
/Users/max/.jbang/cache/urls/b417f02218eb9188e09a82b92fbf3634b9648f055f62b8d95e35f58682651d10/helloworld.groovy.java:2: error: class, interface, or enum expected
println("Hello Groovy!")
^
1 error
[jbang] [ERROR] Error during compile
[jbang] Run with --verbose for more details
```
so seems like something is incomplete in the fetching/identification of groovy.

**Comment by linux-china** [comment]
It's a bug?  and I posted  it on  #1137

File: src/main/java/dev/jbang/source/GroovyScriptSource.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+		super(script, replaceProperties);
+	}
+
+	@Override
+	public List<String> getCompileOptions() {
+		return Collections.emptyList();
+	}
+
+	public List<String> getRuntimeOptions() {
+		return Arrays.asList("-Dgroovy.grape.enable=false");
```
[the remark this answers is not in the document]
Comment (linux-china) [reply src/main/java/dev/jbang/source/GroovyScriptSource.java:28]: yes. Disable grape to resolve dependencies  dynamically because dependencies has been solved by JBang.

File: src/main/java/dev/jbang/cli/BaseBuildCommand.java [hunk 8]
```diff
@@ -170,7 +168,12 @@ public static IntegrationResult buildJar(ScriptSource src, RunContext ctx, File
 		Util.infoMsg("Building jar...");
 		Util.verboseMsg("compile: " + String.join(" ", optionList));
 
-		Process process = new ProcessBuilder(optionList).inheritIO().start();
+		final ProcessBuilder processBuilder = new ProcessBuilder(optionList).inheritIO();
+		if (src instanceof GroovyScriptSource) {
```
Comment (maxandersen) [inline src/main/java/dev/jbang/cli/BaseBuildCommand.java:172]: why the need for changing JAVA_HOME ?

and then why remove GROOY_HOME?

Comment (linux-china) [reply src/main/java/dev/jbang/cli/BaseBuildCommand.java]: Code snippet from startGroovy(called by groovyc):  

```bash
# For Cygwin, ensure paths are in UNIX format before anything is touched.
if $cygwin ; then
    [ -n "$GROOVY_HOME" ] && GROOVY_HOME="$(cygpath --unix "$GROOVY_HOME")"
    [ -n "$JAVACMD" ] && JAVACMD="$(cygpath --unix "$JAVACMD")"
    [ -n "$JAVA_HOME" ] && JAVA_HOME="$(cygpath --unix "$JAVA_HOME")"
    [ -n "$CP" ] && CP=$(cygpath --path --unix "$CP")
else
    if [ -n "$GROOVY_HOME" ] && [ "$(expr "$GROOVY_HOME" : '\/$')" ] ; then
        GROOVY_HOME="$(echo "$GROOVY_HOME" | sed -e 's/\/$//')"
    fi
fi
```

If developer has defined JAVA_HOME to 8 and GROOVY_HOME to 2.5.6 already,  but the Groovy script use `//JAVA 17` and `//GROOVY 3.0.9`,  and groovyc could not compile groovy code to class correctly because groovyc will use env variables.  For examples:

* Script uses API from Java 17 only
* Script uses API from Groovy 3.0 only

So the solution is to reset JAVA_HOME and remove GROOVY_HOME.  It's ok without GROOVY_HOME env, and groovyc will use its parent directory as GROOVY_HOME.

Comment (maxandersen) [reply src/main/java/dev/jbang/cli/BaseBuildCommand.java]: Just that we do set java home in some cases I'm not fond of the imbalance. But since this code at the moment only affect groovy I merged it and we iterate on it.
